File: sdk/src/main/common/HttpClient.java
package common;

import android.app.Application;

import java.net.CookieManager;
import java.net.CookiePolicy;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import okhttp3.CookieJar;
import okhttp3.Interceptor;
import okhttp3.JavaNetCookieJar;
import okhttp3.logging.HttpLoggingInterceptor;

public class HttpClient {

    private static Map<String, RetrofitHttpClient> clientMap = new HashMap<>();

    static CookieManager cookieManager;

    static CookieJar cookieJar;
    static HttpLoggingInterceptor.Level httpLoggingInterceptor = HttpLoggingInterceptor.Level.NONE;

    public static void setHttpLoggingInterceptor(HttpLoggingInterceptor.Level httpLoggingInterceptor) {
        HttpClient.httpLoggingInterceptor = httpLoggingInterceptor;
    }

    public static RetrofitHttpClient initialize
            (Application app, String baseUrl, Map<String, String> headers,
             Map<String, List<Interceptor>> interceptorList, String namespace) {
        if (null == cookieManager) {
            cookieManager = new CookieManager(new PersistentCookieStore
                    (app.getApplicationContext()), CookiePolicy.ACCEPT_ALL);
            cookieJar = new JavaNetCookieJar(cookieManager);
        }

        if (null == clientMap.get(namespace)) {
            RetrofitHttpClient retrofitHttpClient = new RetrofitHttpClient(app, baseUrl,
                    headers, interceptorList, cookieJar, cookieManager);

            clientMap.put(namespace, retrofitHttpClient);
        }
        return clientMap.get(namespace);
    }
}

File: src/main/java/com/sdk/common/HttpClient.kt
package com.sdk.common

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import okhttp3.CookieJar
import okhttp3.Interceptor
import okhttp3.JavaNetCookieJar
import okhttp3.logging.HttpLoggingInterceptor
import java.net.CookieManager
import java.net.CookiePolicy
import java.net.CookieStore
import java.util.*

object HttpClient {
    private val clientMap: MutableMap<String, RetrofitHttpClient?> = HashMap()
    var cookieManager: CookieManager? = null
    var cookieJar: CookieJar? = null
    var httpsLoggingInterceptor = HttpLoggingInterceptor.Level.NONE
    private var currentDomainUrl = ""
    val gson: Gson by lazy { getGsonInstance() }

    private fun getGsonInstance(): Gson {
        return GsonBuilder()
            .registerTypeAdapterFactory(IgnoreJsonParsingExceptionsTypeAdapterFactory())
            .setLenient()
            .create()
    }

    fun setHttpLoggingInterceptor(httpLoggingInterceptor: HttpLoggingInterceptor.Level) {
        this.httpsLoggingInterceptor = httpLoggingInterceptor
    }

    fun getHttpLoggingInterceptor(): HttpLoggingInterceptor.Level {
        return httpsLoggingInterceptor
    }

    fun initialize(
        baseUrl: String,
        headers: Map<String, String>? = null,
        interceptorList: Map<String, List<Interceptor>>? = null,
        namespace: String = "client",
        persistentCookieStore: CookieStore? = null,
        certPublicKey: String? = null
    ): RetrofitHttpClient? {
        if (persistentCookieStore != null) {
            cookieManager = CookieManager(persistentCookieStore, CookiePolicy.ACCEPT_ALL)
            cookieManager?.let {
                cookieJar = JavaNetCookieJar(it)
            }
        }
        if (!currentDomainUrl.equals(baseUrl, ignoreCase = true)) {
            clientMap.clear()
            currentDomainUrl = baseUrl
        }
        if (null == clientMap[namespace]) {
            val retrofitHttpClient = RetrofitHttpClient(
                baseUrl,
                headers, interceptorList, cookieJar, cookieManager, persistentCookieStore, certPublicKey
            )
            clientMap[namespace] = retrofitHttpClient
        }
        return clientMap[namespace]
    }

    fun setDebuggable(debuggable: Boolean) {
        httpsLoggingInterceptor = if (debuggable) {
            HttpLoggingInterceptor.Level.BODY
        } else {
            HttpLoggingInterceptor.Level.NONE
        }
    }

    fun reset() {
        clientMap.clear()
    }

}